llvm/lib/Transforms/IPO/OpenMPOpt.cpp

```diff
@@ -2621,6 +2621,17 @@ struct AAICVTrackerCallSiteReturned : AAICVTracker {
   }
 };
 
+/// Determines if \p BB exits the function unconditionally itself or reaches a
+/// block that does through only unique successors.
+static bool hasFunctionEndAsUniqueSuccessor(const BasicBlock *BB) {
+  if (succ_empty(BB))
+    return true;
+  const BasicBlock *const Successor = BB->getUniqueSuccessor();
+  if (!Successor)
+    return false;
+  return hasFunctionEndAsUniqueSuccessor(Successor);
```

Comment:
It is likely not to matter here but we usually avoid recursion and go with loops.
Also, const BB * const is a long way to spell auto.

Reply:
Yes, this is tail recursion so it should be functionally equivalent